packages/store/src/store.ts

```diff
@@ -84,7 +98,50 @@ export class Store {
    */
   selectSignal<T>(selector: TypedSelector<T>): Signal<T> {
     const selectorFn = this.getStoreBoundSelectorFn(selector);
-    return computed<T>(() => selectorFn(this._stateStream.state()));
+    const initialValue = selectorFn(this._stateStream.value);
+    const observable = this.selectFromStateStream(
+      // https://github.com/ngxs/store/issues/2180
+      // This is explicitly piped with the `debounceTime` to prevent synchronous
+      // signal updates. Signal updates occurring within effects can lead to errors
+      // stating that signal writes are not permitted in effects. This approach helps
+      // decouple signal updates from synchronous changes, ensuring compliance with
+      // constraints on updates inside effects.
+      // Developers should never rely on manually reading the signal after the state
+      // has been updated, whether synchronously or asynchronously, since the expected
+      // result may not be immediately available. To retrieve the current slice of the
+      // state, use `selectSnapshot` instead of directly accessing the signal. Signals
+      // are intended for use in templates or effects, as they always guarantee
+      // consistency with the latest signal value.
+      this._selectableStateStream.pipe(debounceTime(0, asapScheduler)),
+      selectorFn
+    );
+
+    try {
+      assertInInjectionContext(() => {});
+      return toSignal(observable, { initialValue });
+    } catch {
+      return toSignal(observable, { initialValue, injector: this._injector });
+    }
```

Comment:
@markwhitfeld I don't think it's the right approach to do this. Basically, we're asserting the injection context, if the `assertInInjectionContext` doesn't throw an error, we're calling `toSignal` inside an injection context. Otherwise, we provide an explicit injector.

Reply:
Well, I actually would think that signals by default would require an injection context (not an explicit injector).

Reply:
I don't think that an injection context is necessary for the signals, but it would definitely be necessary for `toSignal` so that the `DestroyRef` can be used for unsubscribing.